### PR TITLE
Implementing com_contact frontend editing

### PR DESCRIPTION
**Pull request by Bakual**
This PR adds frontend editing to com_contact.

### Summary of Changes
Adding form view and accompagning controller and model to com_contact frontend

### Testing Instructions
* Test adding and editing contacts from frontend. If the user is authorised, he should see an edit link in the category view and contact view. The category view additionaly has a "New" button.
The featured view doesn't have any such links.
There is a menuitem to point to create new contacts.

* Test that data doesn't get lost when saving (especially data not available to edit in frontend vs backend).

* Test especially also ACL. JSST should look over it for possible security issues.

### Expected result
Works

### Actual result
Frontend editing not possible.

### Documentation Changes Required
A new help page will be needed. Possibly also other doc pages.

### CODE REVIEW

File: components/com_contact/helpers/icon.php [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+defined('_JEXEC') or die;
+
+use Joomla\CMS\Language\Text;
+use Joomla\Registry\Registry;
+
+/**
+ * Contact Component HTML Helper
+ *
+ * @since  1.5
```
Comment (brianteeman) [inline components/com_contact/helpers/icon.php]: ??

File: language/en-GB/en-GB.com_contact.ini [hunk 2]
```diff
@@ -40,13 +44,58 @@ COM_CONTACT_ENQUIRY_TEXT="This is an enquiry email via %s from:"
 COM_CONTACT_ERROR_CONTACT_NOT_FOUND="Contact not found"
 COM_CONTACT_FAX="Fax"
 COM_CONTACT_FAX_NUMBER="Fax: %s"
+COM_CONTACT_FIELD_CREATED_BY_ALIAS_DESC="Enter an alias to be displayed instead of the name of the user who created the contact."
+COM_CONTACT_FIELD_CREATED_BY_ALIAS_LABEL="Created By Alias"
+COM_CONTACT_FIELD_FEATURED_DESC="If marked yes, will be displayed in featured view."
```
Comment (brianteeman) [inline language/en-GB/en-GB.com_contact.ini:49]: Is there no label?

File: components/com_contact/helpers/icon.php (hunk 1 above)
Comment (Bakual) [reply components/com_contact/helpers/icon.php]: Copy-Paste error :)
I've based much of the code on com_content and adjusted/removed the code. This one sneaked through.

File: language/en-GB/en-GB.com_contact.ini (hunk 2 above)
Comment (Bakual) [reply language/en-GB/en-GB.com_contact.ini:49]: The label is a global one: `JFEATURED="Featured"`. Same as in backend.

**Comment by Bakual** [comment]
I will solve the various codestyle issues later. Those are the ones PhpStorm either does wrong or doesn't detect.

**Comment by Bakual** [comment]
Codestyle issues solved now.

**Comment by Quy** [comment]
After saving in PHP error log:
> PHP Notice:  Undefined index: params in \administrator\components\com_contact\models\contact.php on line 353

File: components/com_contact/views/category/tmpl/default_items.php [hunk 3]
```diff
@@ -21,17 +24,17 @@
...
 					<input
-						type="text"
-						name="filter-search"
-						id="filter-search"
-						value="<?php echo $this->escape($this->state->get('list.filter')); ?>"
-						class="inputbox"
-						onchange="document.adminForm.submit();"
-						title="<?php echo JText::_('COM_CONTACT_FILTER_SEARCH_DESC'); ?>"
-						placeholder="<?php echo JText::_('COM_CONTACT_FILTER_SEARCH_DESC'); ?>"
+							type="text"
```
Comment (Quy) [inline components/com_contact/views/category/tmpl/default_items.php]: Revert indentation.

Comment (Bakual) [reply components/com_contact/views/category/tmpl/default_items.php]: Done

**Comment by Bakual** [comment]
@Quy Should be fine now. The notice happend because the backend model expected the data to be present, but in frontend the links aren't in the form.

File: administrator/components/com_contact/models/contact.php [hunk 4]
```diff
@@ -57,6 +57,15 @@ class ContactModelContact extends JModelAdmin
 		'user_id'       => 'batchUser',
 	);
 
+	/**
+	 * Name of the form
+	 *
+	 * @var string
+	 *
```
Comment (Quy) [inline administrator/components/com_contact/models/contact.php]:
```suggestion

```

**Comment by Quy** [comment]
Missing `COM_CONTACT_FIELD_LANGUAGE_DESC`

language

File: components/com_contact/controllers/contact.php [hunk 5]
```diff
@@ -266,4 +286,147 @@ private function _sendEmail($data, $contact, $copy_email_activated)
...
+	{
+		$return = $this->input->get('return', null, 'base64');
+
+		if (empty($return) || !Uri::isInternal(base64_decode($return)))
+		{
+			return Uri::base();
+		}
+		else
+		{
+			return base64_decode($return);
```
Comment (Quy) [inline components/com_contact/controllers/contact.php]: Very minor suggestion to move the `return` and remove the `else` statement.

File: components/com_contact/models/form.php [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+ *
+ * @since  __DEPLOY_VERSION__
+ */
+class ContactModelForm extends ContactModelContact
+{
+	/**
+	 * Model typeAlias string. Used for version history.
+	 *
+	 * @var  string
+	 *
```
Comment (Quy) [inline components/com_contact/models/form.php]:
```suggestion

```

File: components/com_contact/models/form.php [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+	 * @since  __DEPLOY_VERSION__
+	 */
+	public $typeAlias = 'com_contact.contact';
+
+	/**
+	 * Name of the form
+	 *
+	 * @var string
+	 *
```
Comment (Quy) [inline components/com_contact/models/form.php]:
```suggestion

```

**Comment by Quy** [comment]
`Version Note (optional)` will not save.

File: components/com_contact/models/forms/form.xml [hunk 8]
```diff
@@ -1,772 +1,378 @@
...
+		<field
+			name="published"
+			type="list"
+			label="JSTATUS"
+			description="JFIELD_PUBLISHED_DESC"
+			default="1"
+			id="published"
+			class="chzn-color-state"
+			size="1"
+		>
```
Comment (Quy) [inline components/com_contact/models/forms/form.xml]:
```suggestion
			>
```

File: components/com_contact/models/forms/form.xml [hunk 9]
```diff
@@ -1,772 +1,378 @@
...
-			label="CONTACT_EMAIL_SHOW_FORM_LABEL"
-			description="CONTACT_EMAIL_SHOW_FORM_DESC"
-			>
-			<option value="">JGLOBAL_USE_GLOBAL</option>
-			<option value="0">JHIDE</option>
-			<option value="1">JSHOW</option>
+			label="JFEATURED"
+			description="COM_CONTACT_FIELD_FEATURED_DESC"
+			default="0"
+		>
```
Comment (Quy) [inline components/com_contact/models/forms/form.xml]:
```suggestion
			>
```

File: components/com_contact/models/forms/form.xml [hunk 10]
```diff
@@ -1,772 +1,378 @@
...
+	<fieldset
+		name="language"
+		label="JFIELD_LANGUAGE_LABEL"
+	>
+		<field
+			name="language"
+			type="contentlanguage"
+			label="JFIELD_LANGUAGE_LABEL"
+			description="COM_CONTACT_FIELD_LANGUAGE_DESC"
+		>
```
Comment (Quy) [inline components/com_contact/models/forms/form.xml]:
```suggestion
			>
```

**Comment by Bakual** [comment]
@Quy Thanks! I've added the missing language string and removed the superfluous `else`.

File: components/com_contact/views/form/view.html.php [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+		if ($menu)
+		{
+			$this->params->def('page_heading', $this->params->get('page_title', $menu->title));
+		}
+		else
+		{
+			$this->params->def('page_heading', Text::_('COM_CONTACT_FORM_EDIT_CONTACT'));
+		}
+
+		$title = $this->params->def('page_title', Text::_('COM_CONTACT_FORM_EDIT_CONTAC'));
```
Comment (Quy) [inline components/com_contact/views/form/view.html.php]:
```suggestion
		$title = $this->params->def('page_title', Text::_('COM_CONTACT_FORM_EDIT_CONTACT'));
```

Comment (Quy) [reply components/com_contact/views/form/view.html.php]: Missing `COM_CONTACT_FORM_EDIT_CONTACT`?

File: components/com_contact/views/form/tmpl/edit.php [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+		</fieldset>
+		<div class="btn-toolbar">
+			<div class="btn-group">
+				<button type="button" class="btn btn-primary" onclick="Joomla.submitbutton('contact.save')">
+					<span class="icon-ok"></span><?php echo Text::_('JSAVE') ?>
+				</button>
+			</div>
+			<div class="btn-group">
+				<button type="button" class="btn" onclick="Joomla.submitbutton('contact.cancel')">
+					<span class="icon-cancel"></span><?php echo Text::_('JCANCEL') ?>
```
Comment (Quy) [inline components/com_contact/views/form/tmpl/edit.php]:
```suggestion
					<span class="icon-cancel" aria-hidden="true"></span><?php echo Text::_('JCANCEL'); ?>
```

File: components/com_contact/views/form/tmpl/edit.php [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+			<?php echo HTMLHelper::_('bootstrap.endTabSet'); ?>
+
+			<input type="hidden" name="task" value=""/>
+			<input type="hidden" name="return" value="<?php echo $this->return_page; ?>"/>
+			<?php echo HTMLHelper::_('form.token'); ?>
+		</fieldset>
+		<div class="btn-toolbar">
+			<div class="btn-group">
+				<button type="button" class="btn btn-primary" onclick="Joomla.submitbutton('contact.save')">
+					<span class="icon-ok"></span><?php echo Text::_('JSAVE') ?>
```
Comment (Quy) [inline components/com_contact/views/form/tmpl/edit.php]:
```suggestion
					<span class="icon-ok" aria-hidden="true"></span><?php echo Text::_('JSAVE'); ?>
```

File: components/com_contact/controllers/contact.php [hunk 14]
```diff
@@ -266,4 +286,145 @@ private function _sendEmail($data, $contact, $copy_email_activated)
...
+				return ($record->created_by == $user->id);
+			}
+		}
+		else
+		{
+			// Since there is no asset tracking, revert to the component permissions.
+			return parent::allowEdit($data, $key);
+		}
+
+		return false;
```
Comment (Quy) [inline components/com_contact/controllers/contact.php]: Move after line 350 and remove superfluous `else`.

Comment (Bakual) [reply components/com_contact/controllers/contact.php]: True, that works.

File: components/com_contact/views/form/view.html.php (hunk 11 above)
Comment (Bakual) [reply components/com_contact/views/form/view.html.php]: Added that string now.

**Comment by BertaOctech** [comment]
I have tested this item :white_check_mark: successfully on 9d1bd59ae2fdc08979850fd5df936fc3fe61bac4<br /><br />We have modify a single contact and it has worked fine
We have created a new contact from the category view and modify it 
We have checked the public cannot edit the contacts
We have created a menu item for new contacts

Everything worked as expected<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22735">issues.joomla.org/tracker/joomla-cms/22735</a>.</sub>

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 9d1bd59ae2fdc08979850fd5df936fc3fe61bac4<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22735">issues.joomla.org/tracker/joomla-cms/22735</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22735">issues.joomla.org/tracker/joomla-cms/22735</a>.</sub>

**Comment by Bakual** [comment]
@joomla/cms-maintainers Can I have a decision here if this goes into 3.10 or 4.0?

**Comment by rdeutz** [comment]
I think the decision was no new features in 3.10 (or whatever the bridge version to 4 is called) so this has to go into 4 if we don't have a minor version before the bridge

**Comment by kofaysi** [comment]
I have tested this item :red_circle: unsuccessfully on 9d1bd59ae2fdc08979850fd5df936fc3fe61bac4<br /><br />All the following are frontend issues.

Custom field groups are not visible and editable. Even not for a new contact.

I sometimes experience a major delay in applying the access or view rights:
* When a non-editor user is logged in after an editor user, the non-editor could see the "New" button, but "Error: Create record not permitted." is displayed on access. The item has been checked-in already by then.
* When an editor user is logged in after a non-editor user, the editor could not see the "New" button. Refresh of the page and Cache clearing in the browser did not help. I had to logout and login and refresh page. Strange.

The privacy request label is not rendered, but I guess this is not belonging to this PR.
* https://screenshots.firefox.com/oV8W4MHpzxwFFdvJ/svjsvetova523-1.cz
* https://screenshots.firefox.com/7lZ0uKzRwc1hnSOC/svjsvetova523-1.cz

In the list of the items (contacts) of the category, two edit link are accessible: an icon and a text link.
* The text link context help (on cursor hover) reads "Edit article".
* The icon context help reads "Edit contact | Published | [some date] | Written by" -- the message is truncated, the name of the editor is not displayed.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22735">issues.joomla.org/tracker/joomla-cms/22735</a>.</sub>

**Comment by kofaysi** [comment]
Oh, and after reverting the patch I realized that the subcategories heading was not displayed in any of the categories, while the patch was applied.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22735">issues.joomla.org/tracker/joomla-cms/22735</a>.</sub>

**Comment by infograf768** [comment]
@Bakual  ?

**Comment by Bakual** [comment]
> Oh, and after reverting the patch I realized that the subcategories heading was not displayed in any of the categories, while the patch was applied.

I can't imagine how that could be related to this PR. I didn't touch the code related to showing subcategories.

> The text link context help (on cursor hover) reads "Edit article".

Found the culprit for that one. Good catch.

> The icon context help reads "Edit contact | Published | [some date] | Written by" -- the message is truncated, the name of the editor is not displayed.

this shows fine for me. Could it be that the user who created that article has been deleted?

Regarding the privacy stuff, I have no clue how that is supposed to work. I've never looked into that so far.

With Editor user you refer to a user with edit privileges? And non-editor would be one who is only able to create?

**Comment by kofaysi** [comment]
> > Oh, and after reverting the patch I realized that the subcategories heading was not displayed in any of the categories, while the patch was applied.
> 
> I can't imagine how that could be related to this PR. I didn't touch the code related to showing subcategories.

I have missed that one. It is me, who did not checked back on it. Sorry, The subcategories show up.

> > The icon context help reads "Edit contact | Published | [some date] | Written by" -- the message is truncated, the name of the editor is not displayed.
> 
> this shows fine for me. Could it be that the user who created that article has been deleted?

No, he/she still exist. I checked and it does not happen with the new commit. The snippet seems non-truncated and full.

> Regarding the privacy stuff, I have no clue how that is supposed to work. I've never looked into that so far.
> 
> With Editor user you refer to a user with edit privileges? And non-editor would be one who is only able to create?

Yes, please. It is about the right to edit and create the contacts. But we have only users who can both create and edit and those who cannot do it.

**Comment by kofaysi** [comment]
I have tested this item :red_circle: unsuccessfully on 4d758cecc37d249ebc611b147e3b953667d0654b<br /><br />I sometimes experience a major delay in applying the access or view rights, described earlier. Login and logout various test users. The edit links appears after few seconds of active browsing the contact items.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22735">issues.joomla.org/tracker/joomla-cms/22735</a>.</sub>

**Comment by kofaysi** [comment]
Custom field groups (tabs) are missing from the edit interface entirely.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22735">issues.joomla.org/tracker/joomla-cms/22735</a>.</sub>

**Comment by Bakual** [comment]
> No, he/she still exist. I checked and it does not happen with the new commit. The snippet seems non-truncated and full.

I haven't changed anything there, so it seems to have been a glitch.

> I sometimes experience a major delay in applying the access or view rights, described earlier. Login and logout various test users. The edit links appears after few seconds of active browsing the contact items.

That's real strange. it's probably related to session lifetimes, cache or whatever. You probably get the same effects with com_content frontend edit.
I doubt it's related to this PR itself.

> Custom field groups (tabs) are missing from the edit interface entirely.

Yep, confirming that and I need to investigate why that is.

I also have to test again with various users to see what this permission thing is.

**Comment by infograf768** [comment]
Back to pending.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22735">issues.joomla.org/tracker/joomla-cms/22735</a>.</sub>

**Comment by Bakual** [comment]
I found the issue with the custom fields. They were actually showing, but those from the mail section and not the ones from the contacts.
I've now renamed the form so I can properly detect it in the validateSection function. Now it works for me.

I couldn't reproduce the permission issue with the editor and non-editor users. That sounds like something on your setup which doesn't play nice. Maybe cache related?

**Comment by kofaysi** [comment]
> 
> 
> I found the issue with the custom fields. They were actually showing, but those from the mail section and not the ones from the contacts.
> I've now renamed the form so I can properly detect it in the validateSection function. Now it works for me.
> 
> I couldn't reproduce the permission issue with the editor and non-editor users. That sounds like something on your setup which doesn't play nice. Maybe cache related?

I switched off System cache (OFF - Caching disabled) and the behavior is correct -- tried to log in and out of various users a couple of times. Previously the progressive caching has been set for system cache. Does that make sense to you? I have no knowledge about these cache relations.

**Comment by kofaysi** [comment]
I have tested this item :white_check_mark: successfully on 3d5c8fa99392bf4c851bcc9f28617e207c011dab<br /><br />Saving data ok.
Custom fields displayed, editable.
The previously experienced regression with view and edit rights was due to system cache probably. System cache set to "OFF - caching disabled" instead of the previous "ON - progressive caching".<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22735">issues.joomla.org/tracker/joomla-cms/22735</a>.</sub>

**Comment by kofaysi** [comment]
By using the frontend, I was able to create two test contacts with the same alias and no warning has been issued.
image

When creating the same test contact in the backend, a warning has been issued
image
Although, no alias has been inserted by the editor, only the contact name.

I have a single remark to both above behaviors:
* The alias should be altered automatically by concatonating "-2", "-3", etc. (as "Save as new" does)

**Comment by Bakual** [comment]
I'll check the cache thing. I haven't used cache much myself but I think it usually gets disabled when a user is logge din. Maybe that's not done in com_contact.
I'll check the alias as well and see how that is done in com_content. So we have the same behaviour in both places.

**Comment by Bakual** [comment]
Closing in favor of rebased 4.0 PR: https://github.com/joomla/joomla-cms/pull/24311

**Pull request closed without merging**